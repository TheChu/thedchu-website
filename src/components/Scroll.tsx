import React, {
  cloneElement,
  Component,
  KeyboardEvent,
  MouseEvent,
  ReactElement,
} from 'react';
import smoothscroll from 'smoothscroll-polyfill';

const Element = (props) => props.children;

interface ScrollProps {
  children: ReactElement;
  element?: string;
  offset?: number;
  onClick?: (event: KeyboardEvent | MouseEvent) => void;
  timeout?: number;
  type?: string;
}

class Scroll extends Component<ScrollProps> {
  constructor(props: ScrollProps) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  componentDidMount(): void {
    smoothscroll.polyfill();
  }

  handleClick = (event: KeyboardEvent | MouseEvent): void => {
    event.preventDefault();
    const { onClick = () => {} } = this.props; // eslint-disable-line
    let elem: Element;
    let scroll = true;
    const { type, element, offset, timeout } = this.props;
    if (type && element) {
      switch (type) {
        case 'class':
          [elem] = document.getElementsByClassName(element);
          scroll = !!elem;
          break;
        case 'id':
          elem = document.getElementById(element);
          scroll = !!elem;
          break;
        default:
      }
    }

    if (scroll) {
      this.scrollTo(elem, offset, timeout);
    } else {
      console.log(`Element not found: ${element}`); // eslint-disable-line
    }

    onClick(event);
  };

  scrollTo = (element: Element, offSet = 0, timeout = null): void => {
    const elemPos = element
      ? element.getBoundingClientRect().top + window.pageYOffset
      : 0;
    if (timeout) {
      setTimeout(() => {
        window.scroll({ top: elemPos + offSet, left: 0, behavior: 'smooth' });
      }, timeout);
    } else {
      window.scroll({ top: elemPos + offSet, left: 0, behavior: 'smooth' });
    }
  };

  render(): ReactElement {
    const { children } = this.props;

    return (
      <Element>
        {cloneElement(children, {
          onClick: this.handleClick,
        })}
      </Element>
    );
  }
}

export default Scroll;
