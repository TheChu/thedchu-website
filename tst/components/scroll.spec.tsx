import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import Scroll from '../../src/components/Scroll';

window.scroll = jest.fn();
Element.prototype.getBoundingClientRect = jest.fn();
(Element.prototype.getBoundingClientRect as jest.Mock).mockReturnValue({
  top: 100,
});
console.log = jest.fn(); // eslint-disable-line

describe('Scroll', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('no props', () => {
    const scrollButtonNoProps = (
      <Scroll>
        <button type="button" aria-label="Test button" />
      </Scroll>
    );

    it('should match Scroll snapshot', () => {
      const { asFragment } = render(scrollButtonNoProps);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should call window.scroll with top as 0', () => {
      const { getByRole } = render(scrollButtonNoProps);
      expect(window.scroll).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).toHaveBeenCalledWith({
        top: 0,
        left: 0,
        behavior: 'smooth',
      });
      expect(Element.prototype.getBoundingClientRect).not.toHaveBeenCalled();
    });
  });

  describe('id type', () => {
    it('should call window.scroll', () => {
      const { getByRole } = render(
        <>
          <Scroll type="id" element="test-div">
            <button type="button" aria-label="Test button" />
          </Scroll>
          <div id="test-div" />
        </>
      );
      expect(window.scroll).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).toHaveBeenCalledTimes(1);
      expect(window.scroll).toHaveBeenCalledWith({
        top: 100,
        left: 0,
        behavior: 'smooth',
      });
    });

    it('should not call window.scroll if element does not exist', () => {
      const { getByRole } = render(
        <>
          <Scroll type="id" element="wrong-div">
            <button type="button" aria-label="Test button" />
          </Scroll>
          <div id="test-div" />
        </>
      );
      expect(window.scroll).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).not.toHaveBeenCalled();
      expect(console.log).toHaveBeenCalledWith('Element not found: wrong-div'); // eslint-disable-line
    });

    describe('offset', () => {
      it('should call window.scroll with offset', () => {
        const { getByRole } = render(
          <>
            <Scroll type="id" element="test-div" offset={50}>
              <button type="button" aria-label="Test button" />
            </Scroll>
            <div id="test-div" />
          </>
        );
        expect(window.scroll).not.toHaveBeenCalled();

        fireEvent.click(getByRole('button', { name: 'Test button' }));
        expect(window.scroll).toHaveBeenCalledTimes(1);
        expect(window.scroll).toHaveBeenCalledWith({
          top: 150,
          left: 0,
          behavior: 'smooth',
        });
      });
    });

    describe('timeout', () => {
      it('should call window.scroll after timeout', async () => {
        const { getByRole } = render(
          <>
            <Scroll type="id" element="test-div" timeout={100}>
              <button type="button" aria-label="Test button" />
            </Scroll>
            <div id="test-div" />
          </>
        );
        expect(window.scroll).not.toHaveBeenCalled();

        fireEvent.click(getByRole('button', { name: 'Test button' }));
        await waitFor(() => {
          expect(window.scroll).toHaveBeenCalledTimes(1);
          expect(window.scroll).toHaveBeenCalledWith({
            top: 100,
            left: 0,
            behavior: 'smooth',
          });
        });
      });
    });

    describe('onClick', () => {
      const mockFunc = jest.fn();

      it('should call window.scroll and the provided onClick function', () => {
        const { getByRole } = render(
          <>
            <Scroll type="id" element="test-div" onClick={mockFunc}>
              <button type="button" aria-label="Test button" />
            </Scroll>
            <div id="test-div" />
          </>
        );
        expect(window.scroll).not.toHaveBeenCalled();
        expect(mockFunc).not.toHaveBeenCalled();

        fireEvent.click(getByRole('button', { name: 'Test button' }));
        expect(window.scroll).toHaveBeenCalledTimes(1);
        expect(window.scroll).toHaveBeenCalledWith({
          top: 100,
          left: 0,
          behavior: 'smooth',
        });
        expect(mockFunc).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('class type', () => {
    it('should call window.scroll', () => {
      const { getByRole } = render(
        <>
          <Scroll type="class" element="test-div">
            <button type="button" aria-label="Test button" />
          </Scroll>
          <div className="test-div" />
        </>
      );
      expect(window.scroll).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).toHaveBeenCalledTimes(1);
      expect(window.scroll).toHaveBeenCalledWith({
        top: 100,
        left: 0,
        behavior: 'smooth',
      });
    });

    it('should not call window.scroll if element does not exist', () => {
      const { getByRole } = render(
        <>
          <Scroll type="class" element="wrong-div">
            <button type="button" aria-label="Test button" />
          </Scroll>
          <div className="test-div" />
        </>
      );
      expect(window.scroll).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).not.toHaveBeenCalled();
      expect(console.log).toHaveBeenCalledWith('Element not found: wrong-div'); // eslint-disable-line
    });

    describe('offset', () => {
      it('should call window.scroll with offset', () => {
        const { getByRole } = render(
          <>
            <Scroll type="class" element="test-div" offset={50}>
              <button type="button" aria-label="Test button" />
            </Scroll>
            <div className="test-div" />
          </>
        );
        expect(window.scroll).not.toHaveBeenCalled();

        fireEvent.click(getByRole('button', { name: 'Test button' }));
        expect(window.scroll).toHaveBeenCalledTimes(1);
        expect(window.scroll).toHaveBeenCalledWith({
          top: 150,
          left: 0,
          behavior: 'smooth',
        });
      });
    });

    describe('timeout', () => {
      it('should call window.scroll after timeout', async () => {
        const { getByRole } = render(
          <>
            <Scroll type="class" element="test-div" timeout={100}>
              <button type="button" aria-label="Test button" />
            </Scroll>
            <div className="test-div" />
          </>
        );
        expect(window.scroll).not.toHaveBeenCalled();

        fireEvent.click(getByRole('button', { name: 'Test button' }));
        await waitFor(() => {
          expect(window.scroll).toHaveBeenCalledTimes(1);
          expect(window.scroll).toHaveBeenCalledWith({
            top: 100,
            left: 0,
            behavior: 'smooth',
          });
        });
      });
    });

    describe('onClick', () => {
      const mockFunc = jest.fn();

      it('should call window.scroll and the provided onClick function', () => {
        const { getByRole } = render(
          <>
            <Scroll type="id" element="test-div" onClick={mockFunc}>
              <button type="button" aria-label="Test button" />
            </Scroll>
            <div id="test-div" />
          </>
        );
        expect(window.scroll).not.toHaveBeenCalled();
        expect(mockFunc).not.toHaveBeenCalled();

        fireEvent.click(getByRole('button', { name: 'Test button' }));
        expect(window.scroll).toHaveBeenCalledTimes(1);
        expect(window.scroll).toHaveBeenCalledWith({
          top: 100,
          left: 0,
          behavior: 'smooth',
        });
        expect(mockFunc).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('invalid type', () => {
    it('should call window.scroll with top as 0', () => {
      const { getByRole } = render(
        <>
          <Scroll type="invalidType" element="test-div">
            <button type="button" aria-label="Test button" />
          </Scroll>
          <div />
        </>
      );
      expect(window.scroll).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).toHaveBeenCalledWith({
        top: 0,
        left: 0,
        behavior: 'smooth',
      });
      expect(Element.prototype.getBoundingClientRect).not.toHaveBeenCalled();
    });
  });

  describe('only onClick', () => {
    const mockFunc = jest.fn();

    it('should call window.scroll with top as 0 and the provided onClick function', () => {
      const { getByRole } = render(
        <>
          <Scroll onClick={mockFunc}>
            <button type="button" aria-label="Test button" />
          </Scroll>
        </>
      );
      expect(window.scroll).not.toHaveBeenCalled();
      expect(mockFunc).not.toHaveBeenCalled();

      fireEvent.click(getByRole('button', { name: 'Test button' }));
      expect(window.scroll).toHaveBeenCalledTimes(1);
      expect(window.scroll).toHaveBeenCalledWith({
        top: 0,
        left: 0,
        behavior: 'smooth',
      });
      expect(mockFunc).toHaveBeenCalledTimes(1);
    });
  });
});
