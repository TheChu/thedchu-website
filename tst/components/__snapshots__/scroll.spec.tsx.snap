// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll no props should match Scroll snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="Test button"
    type="button"
  />
</DocumentFragment>
`;
